.github/workflows/lintcheck.yml

```diff
@@ -0,0 +1,115 @@
+name: Lintcheck
+
+on: pull_request
+
+env:
+  RUST_BACKTRACE: 1
+  CARGO_INCREMENTAL: 0
+
+concurrency:
+  # For a given workflow, if we push to the same PR, cancel all previous builds on that PR.
+  group: "${{ github.workflow }}-${{ github.event.pull_request.number}}"
+  cancel-in-progress: true
+
+jobs:
+  # Generates `lintcheck-logs/base.json` and stores it in a cache
+  base:
+    runs-on: ubuntu-latest
+
+    outputs:
+      key: ${{ steps.key.outputs.key }}
+
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v4
+      with:
+        fetch-depth: 2
+
+    # HEAD is the generated merge commit `refs/pull/N/merge` between the PR and `master`, `HEAD^`
+    # being the commit from `master` that is the base of the merge
+    - name: Checkout base
+      run: git checkout HEAD^
+
+    # Use the lintcheck from the PR to generate the JSON in case the PR modifies lintcheck in some
+    # way
+    - name: Checkout current lintcheck
+      run: |
+        rm -rf lintcheck
+        git checkout ${{ github.sha }} -- lintcheck
+
+    - name: Create cache key
+      id: key
+      run: echo "key=lintcheck-base-${{ hashfiles('lintcheck/**') }}-$(git rev-parse HEAD)" >> "$GITHUB_OUTPUT"
```

Comment:
Using cache here is a bit weird IMO. I think the `upload-artifacts` action would fit better here? Or am I missing something?

Reply:
It does do some caching as well as act as a way to share it between steps e.g. here `base` [had a cache hit](https://github.com/rust-lang/rust-clippy/actions/runs/9528672639/job/26266611003)

I don't know if you can do that with artifacts, mostly I went with cache because I'm already familiar with it

Reply:
Ah ok. True, `base` can have cache hits between PRs based on the same master commit.

`head` however will never have a cache hit, neither will `diff`. I think the cleaner way would be to just cache `base`. To share the `json` files however, it would be cleaner to use `{down,up}load-artifacts` IMO.

But this is a NIT on the action